Support Slash Commands

JustCat80 opened the pull request:
https://discord.com/developers/docs/interactions/slash-commands

Adds support for general Slash Command things. As for responding to them, check out pr #1210

~~I'll probably wait for #1210 to get pulled before i mark this as ready for review at all~~
## Note:
~~_As per [Abal's comment on slash commands](https://github.com/abalabahaha/eris/issues/1100#issuecomment-812851950): 
"Eris doesn't plan to implement full support for slash commands until the API gets improved+stabilized."_~~

~~While i personally believe its stable enough, its still not very feature filled so it may be a while before this gets pulled.~~

**As per [this discord post](https://support-dev.discord.com/hc/en-us/articles/4404772028055) this is now required soon because as of april 2022 message content will become a privileged intent**
## Changes:
### Client Methods:
- Client#bulkEditCommandPermissions
https://discord.com/developers/docs/interactions/slash-commands#batch-edit-application-command-permissions
- Client#bulkEditCommands
https://discord.com/developers/docs/interactions/slash-commands#bulk-overwrite-global-application-commands
- Client#bulkEditGuildCommands
https://discord.com/developers/docs/interactions/slash-commands#bulk-overwrite-guild-application-commands
- Client#createCommand
https://discord.com/developers/docs/interactions/slash-commands#create-global-application-command
- Client#createGuildCommand
https://discord.com/developers/docs/interactions/slash-commands#create-guild-application-command
- Client#deleteCommand
https://discord.com/developers/docs/interactions/slash-commands#delete-global-application-command
- Client#deleteGuildCommand
https://discord.com/developers/docs/interactions/slash-commands#delete-guild-application-command
- Client#editCommand
https://discord.com/developers/docs/interactions/slash-commands#edit-global-application-command
- Client#editCommandPermissions
https://discord.com/developers/docs/interactions/slash-commands#edit-application-command-permissions
- Client#editGuildCommand
https://discord.com/developers/docs/interactions/slash-commands#edit-guild-application-command
- Client#getCommand
https://discord.com/developers/docs/interactions/slash-commands#create-global-application-command
- Client#getCommandPermissions
https://discord.com/developers/docs/interactions/slash-commands#get-application-command-permissions
- Client#getCommands
https://discord.com/developers/docs/interactions/slash-commands#get-global-application-command
- Client#getGuildCommand
https://discord.com/developers/docs/interactions/slash-commands#get-guild-application-command
- Client#getGuildCommandPermissions
https://discord.com/developers/docs/interactions/slash-commands#get-guild-application-command-permissions
- Client#getGuildCommands
https://discord.com/developers/docs/interactions/slash-commands#get-guild-application-commands
### Guild Methods:
- Guilld#bulkEditCommandPermissions
https://discord.com/developers/docs/interactions/slash-commands#batch-edit-application-command-permissions
- Guild#bulkEditCommands
https://discord.com/developers/docs/interactions/slash-commands#bulk-overwrite-guild-application-commands
- Guild#createCommand
https://discord.com/developers/docs/interactions/slash-commands#create-guild-application-command
- Guild#deleteCommand
https://discord.com/developers/docs/interactions/slash-commands#delete-guild-application-command
- Guild#editCommand
https://discord.com/developers/docs/interactions/slash-commands#edit-guild-application-command
- Guilld#editCommandPermissions
https://discord.com/developers/docs/interactions/slash-commands#edit-application-command-permissions
- Guild#getCommand
https://discord.com/developers/docs/interactions/slash-commands#get-guild-application-command
- Guilld#getCommandPermissions
https://discord.com/developers/docs/interactions/slash-commands#get-application-command-permissions
- Guild#getCommands
https://discord.com/developers/docs/interactions/slash-commands#get-guild-application-commands
- Guilld#getGuildCommandPermissions
https://discord.com/developers/docs/interactions/slash-commands#get-guild-application-command-permissions
### Misc:
- Consistants.CommandOptionTypes
- Consistants.CommandPermissionTypes
- Consistants.CommandTypes
- Slash command endpoints
- ts docs

## TODO: 
### Misc:
- more testing
- make sure context menus work when they release

Commit:
Added endpoints, Added consistents

bsian03 commented on lib/rest/Endpoints.js:
```diff
@@ -36,6 +38,8 @@ module.exports.GUILD_AUDIT_LOGS =                                      (guildID)
 module.exports.GUILD_BANS =                                            (guildID) => `/guilds/${guildID}/bans`;
 module.exports.GUILD_CHANNELS =                                        (guildID) => `/guilds/${guildID}/channels`;
+module.exports.GUILD_COMMAND =               (applicationId, guildId, commandId) => `/applications/${applicationId}/guilds/${guildId}/commands/${commandId}`;
+module.exports.GUILD_COMMANDS =                         (applicationId, guildId) => `/applications/${applicationId}/guilds/${guildId}/commands`;
```
`ID` instead of `Id`

Commit:
Id => ID

Commit:
Add more endpoints

Commit:
Add most client slash command functions

Still missing all Permissions related ones

Commit:
Fixed slash command docs

Commit:
Added most guild slash command functions

JustCat80 committed:
Merge pull request #20 from abalabahaha/dev

Catchup dev updates

Dramex commented on lib/Client.js:
```diff
@@ -478,6 +478,49 @@ class Client extends EventEmitter {
+    }
+
+    /**
+    * Bulk create global slash commands
```
isn't this should be Bulk update or edit global slash commands as in discord docs
https://discord.dev/interactions/slash-commands#bulk-overwrite-global-application-commands

DonovanDMC replied:
It is, and you linked the wrong docs section
https://discord.com/developers/docs/interactions/slash-commands#bulk-overwrite-global-application-commands
https://github.com/Catboi8/eris/blob/slash-commands/lib/rest/Endpoints.js#L10

Dramex replied:
@DonovanDMC my scroller bad, but yeah I don't think createCommands is right.

JustCat80 replied:
Yes in the docs its bulk update but i went the way of deleteMessages and just made it createCommands, tho maybe i should make it createGlobalCommands, also i will change it if enough people think it should be bulkEditCommands

also not only bulk update uses the commands endpoint so that would stay the same

JustCat80 replied:
Also it isnt only for editing commands is the thing, it creates commands and simply overrides existing commands if the names match

Dramex replied:
bulkEditCommands sounds better 👍
because that endpoint updates existing commands too and it doesn't only create

JustCat80 replied:
Alright 3 is enough for me, one min

Commit:
bulkEditCommands

createCommands => bulkEditCommands

createGuildCommands => bulkEditGuildCommands

Commit:
Merge branch 'slash-commands' of https://github.com/Catboi8/eris into slash-commands

JustCat80 committed:
Merge pull request #30 from abalabahaha/dev

Dev updates

Commit:
Ignore .vscode

Commit:
Remove Guild#deleteGuildCommand, fix some docs

Commit:
Fix some docs

Commit:
Added Permission Functions to client and guild

- Client#getGuildCommandPermissions
- Client#getCommandPermissions
- Client#editCommandPermissions
- Client#bulkEditCommandPermissions

- Guilld#getGuildCommandPermissions
- Guilld#getCommandPermissions
- Guilld#editCommandPermissions
- Guilld#bulkEditCommandPermissions

remove trailing spaces

Commit:
Added CommandTypes to Constants and command.type to jsdocs

Note: Currently in private beta

Commit:
Added ts docs :)

Will need to be double checked, I'm not the greatest at ts

Commit:
Oops, linted ts

Commit:
Added NUMBER (type 10) to Constants.CommandOptionTypes

DetachHead commented on index.d.ts:
```diff
@@ -70,6 +70,45 @@ declare namespace Eris {
+    defaultPermission?: boolean;
+  };
+
+  type SlashCommandStructure = Exclude<SlashCommand, "id" | "application_id" | "guild_id">;
```
this should be `Omit` to remove the keys
```suggestion
  type SlashCommandStructure = Omit<SlashCommand, "id" | "application_id" | "guild_id">;
```

JustCat80 committed:
Update index.d.ts

Co-authored-by: DetachHead <57028336+DetachHead@users.noreply.github.com>

DetachHead commented on lib/Client.js:
```diff
@@ -337,6 +337,58 @@ class Client extends EventEmitter {
+    * @returns {Promise<Array>} Resolves with an array of commands objects
+    */
+    bulkEditCommands(commands) {
+        for(const command in commands) {
```
```suggestion
        for(const command of commands) {
```

DetachHead commented on lib/Client.js:
```diff
@@ -337,6 +337,58 @@ class Client extends EventEmitter {
+    * @returns {Promise<Object>} Resolves with an array of commands objects
+    */
+    bulkEditGuildCommands(guildID, commands) {
+        for(const command in commands) {
```
```suggestion
        for(const command of commands) {
```

Commit:
for...in => for...of

Co-Authored-By: DetachHead <57028336+DetachHead@users.noreply.github.com>
